File: A4/polish.pas
program polish;

const  maxin = 40;
       maxout = 40;
       maxstack = 40;
       
type   stores = array[0..maxout] of char;

var    instring: string[maxin];
       outstring, stack: stores;
       operandset, operatorset: set of char;
       out, top, lastout, lastin, item: integer;
       discard: char;
       
procedure push(it: char);
begin
    if top < maxstack
    then begin
        top := top + 1;
        stack[top] := it
    end else begin
        writeln('stack overflow');
        exit()
    end
end;

function pop: char;
begin
    if top > 0
    then begin
        pop := stack[top];
        top := top - 1;
    end else begin
        writeln('stack underflow');
        exit()
    end
end;

procedure move(it: char);
begin
    if it = '('
    then begin
        writeln(' unmatched bracket ');
        exit()
    end else if lastout < maxout
             then begin
                 lastout := lastout + 1;
                 outstring[lastout] := it;
             end else begin
                 writeln(' output string full ');
                 exit()
             end
end;

function priority(it: char): integer;
begin
    case it of
    ')','%' : priority := -1;
        '(' : priority := 0;
    '+','-' : priority := 1;
    '*','/' : priority := 2;
        '^' : priority := 3;
    end
end;

procedure pushORmove(its: char);
var itemtostack: boolean;
begin
    if its = '('
    then push(its)
    else if its = ')'
         then begin
             while priority(stack[top]) <> priority('(') do move(pop);
             discard := pop
         end else if its in operatorset
                  then repeat
                      itemtostack := priority(its) > priority(stack[top]);
                      if itemtostack
                      then push(its)
                      else move(pop)
                  until itemtostack
                  else begin
                      writeln(' invalid operator ', its);
                      exit()
                  end
end;

begin
    lastin := 0;
    operatorset := ['%','+','-','*','/','^'];
    operandset := ['a'..'z','0'..'9'];
    
    while lastin <> 1 do
    begin
        write(' type an expression: ');
        readln(instring);
        lastin := length(instring);
        item := 0;
        lastout := -1;
        stack[0] := '%';
        top := 0;
        
        repeat 
            item := item + 1;
            if instring[item] in operandset
            then move(instring[item])
            else if instring[item] in (operatorset + ['(',')'])
                 then pushORmove(instring[item])
                 else begin
                     writeln(' illegal symbol ', instring[item]);
                     item := lastin;
                 end
        until item = lastin;
        
        while top > 0 do move(pop); {purge stack}
        
        for out := 0 to lastout do write(outstring[out]);
    end
end.
            

File: A4/polish.py
#!/usr/bin/python
#A4: Reverse polish form: Python
#Written by Christopher Katsaras
#April 7th/2017
#
#***********
#Compile/run
#***********
#Type "make python"
#
#*****************
#Known limitations
#*****************
#1.Spaces in inputted equation may cause incorrect output
#e.g 1   + 2 * (3-  2) = NOT SUPPORTED :(
#e.g 1+2*(3-2) = SUPPORTED :)
#!2.Equations should be inputted with NO spaces for best results 

originalString = []
polishString = []
stack = []
playAgain = 'Y'

#Function definitions
#
#Pushes item to the stack
def push(symbol, stack, top):
	top += 1
	try:
		stack[top] = symbol
	except IndexError:
		stack.insert(top,symbol)	
	return top;

#Pops item off stack and put inserts it into the output string
def pop(polishString, stack, top):
	polishString.insert(polishLength,stack[top])
	top -= 1
	return top;	

#Returns the priority of a specific symbol
def priority(sym):

	if sym == '%' or sym == ')':
		return -1;
	elif sym == '(':
		return 0;
	elif sym == '+' or sym == '-':
		return 1;
	elif sym == '*' or sym == '/':
		return 2;
	elif sym == '^':
		return 3;		
	else:
		return -2;

#Loops until user chooses to quit
while (playAgain == 'Y'):
	print "Please input an algebraic expression to convert (No spaces, please)"

	originalString = raw_input()
	length = len(originalString)
	polishLength = 0
	top = 0

	stack.insert(top,'%')

	#Uses a state based parsing method where each character is examined individually 
	for char in originalString:
		if char.isalnum():
			polishLength += 1
			polishString.insert(polishLength,char)
		elif char == '%' or char == '+' or char == '-' or char == '*' or char == '/' or char == '^':
			while (priority(char) <= priority(stack[top])):
				polishLength += 1
				top = pop(polishString,stack,top)
			top = push(char,stack,top)
		elif char == '(':
			top = push(char,stack,top)
		elif char == ')':
			while (priority(stack[top]) != priority('(')):
				polishLength += 1	
				top = pop(polishString,stack,top)
			top -= 1	
		else :
			print "Invalid operator"	

	while (top > 0):
		if(stack[top] == '('):
			print "Unmatched brackets"
		else :
			polishLength += 1
			polishString.insert(polishLength,stack[top])
		top -= 1
	
	print "".join(polishString) #Formats output for the user
	playAgain = raw_input("Would you like to convert another expression? (Type Y to play again or any other key to exit)")
	stack[:] = [] #clears the stack and polishString lists
	polishString[:] = []

print "Goodbye"	
